core/sr-primitives/src/curve.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Provides some utilities to define a piecewise linear function.
+
+use crate::{Perbill, traits::{SimpleArithmetic, SaturatedConversion}};
+use core::ops::Sub;
+
+/// Piecewise Linear function in [0, 1] -> [0, 1].
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(PartialEq, Eq)]
+pub struct PiecewiseLinear<'a> {
+	/// Array of points. Must be in order from the lowest abscissas to the highest.
+	pub points: &'a [(Perbill, Perbill)]
+}
+
+fn abs_sub<N: Ord + Sub<Output=N> + Clone>(a: N, b: N) -> N where {
+	a.clone().max(b.clone()) - a.min(b)
+}
+
+impl<'a> PiecewiseLinear<'a> {
+	/// Compute `f(n/d)*d` with `n <= d`. This is useful to avoid loss of precision.
+	pub fn calculate_for_fraction_times_denominator<N>(&self, n: N, d: N) -> N where
+		N: SimpleArithmetic + Clone
+	{
+		let n = n.min(d.clone());
+
+		if self.points.len() == 0 {
+			return N::zero()
+		}
+
+		let next_point_index = self.points.iter()
+			.position(|p| n < p.0 * d.clone());
+
+		let (prev, next) = if let Some(next_point_index) = next_point_index {
+			if let Some(previous_point_index) = next_point_index.checked_sub(1) {
+				(self.points[previous_point_index], self.points[next_point_index])
+			} else {
+				// There is no previous points, take first point ordinate
+				return self.points.first().map(|p| p.1).unwrap_or_else(Perbill::zero) * d
+			}
+		} else {
+			// There is no next points, take last point ordinate
+			return self.points.last().map(|p| p.1).unwrap_or_else(Perbill::zero) * d
+		};
+
+		let delta_y = multiply_by_rational_saturating(
+			abs_sub(n.clone(), prev.0 * d.clone()),
+			abs_sub(next.1.into_parts(), prev.1.into_parts()),
+			// Must not saturate as prev abscissa > next abscissa
+			next.0.into_parts().saturating_sub(prev.0.into_parts()),
+		);
+
+		// If both substration are same sign then result is positive
+		if (n > prev.0 * d.clone()) == (next.1.into_parts() > prev.1.into_parts()) {
+			(prev.1 * d).saturating_add(delta_y)
+		// Otherwise result is negative
+		} else {
+			(prev.1 * d).saturating_sub(delta_y)
+		}
+	}
+}
+
+// Compute value * p / q.
+// This is guarantee not to overflow on whatever values nor loose precision.
```

Comment:
Should we add this note: `// Mul can saturate if p > q`? meaning that this is 100% safe and accurate given `p < q`

Reply:
At some point it is always 100% safe just when your result doesn't fit into N then it saturates with N::max_value

Reply:
My misunderstanding was due to confusing overflow and precision. makes sense. Nonetheless: 

```suggestion
// This is guaranteed not to overflow on whatever values nor lose precision.
```